File: 763. Partition Labels.py
class Solution:
    def partitionLabels(self, S: str) -> List[int]:
        dic = {}
        for i, char in enumerate(S):
            dic[char] = i
        res = []
        lastpos,i,prev = 0,0,0
        for i, c in enumerate(S):
            lastpos = max(lastpos, dic[c])
            if i == lastpos:
                res.append(i-prev+1)
                prev = i+1
        return res

File: 763. Partition Labels.java
class Solution {
    public List<Integer> partitionLabels(String S) {
        Map<Character, Integer> lastOccur = new HashMap<>();
        for (int i=0; i<S.length(); i++) {
            lastOccur.put(S.charAt(i), i);
        }
        List<Integer> ans = new ArrayList<>();
        int start=0, end=0;
        for (int i=0; i<S.length(); i++) {
            end = Math.max(end, lastOccur.get(S.charAt(i)));
            if (end == i) {
                ans.add(end-start+1);
                start = end+1;
            }
        }
        return ans;
    }
}